test/zdtm/static/pidfd_dead.c

```diff
@@ -0,0 +1,174 @@
+#include <sys/statfs.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+
+#include "zdtmtst.h"
+
+const char *test_doc = "Check C/R of pidfds that point to dead processes\n";
+const char *test_author = "Bhavik Sachdev <b.sachdev1904@gmail.com>";
+
+#ifndef PID_FS_MAGIC
+#define PID_FS_MAGIC 0x50494446
+#endif
+
+static long get_fs_type(int lfd)
+{
+	struct statfs fst;
+
+	if (fstatfs(lfd, &fst)) {
+		return -1;
+	}
+	return fst.f_type;
+}
+
+static int pidfd_open(pid_t pid, unsigned int flags)
+{
+	return syscall(__NR_pidfd_open, pid, flags);
+}
+
+static int pidfd_send_signal(int pidfd, int sig, siginfo_t* info, unsigned int flags)
+{
+	return syscall(__NR_pidfd_send_signal, pidfd, sig, info, flags);
+}
+
+static int open_pidfd_pair(int pidfd[2], int pid)
+{
+	pidfd[0] = pidfd_open(pid, 0);
+	if (pidfd[0] < 0)
+		return 1;
+	pidfd[1] = pidfd_open(pid, 0);
+	if (pidfd[1] < 0)
+		return 1;
+	return 0;
+}
+static int compare_pidfds(int pidfd[2])
+{
+	/*
+	* After linux 6.9 we can compare inode numbers
+	* to determine if two pidfds point to the same process.
+	* While the inode number may change before and after C/R
+	* pidfds pointing to the same pid should have the same inode number.
+	*/
+	struct statx stats[2];
+	statx(pidfd[0], "", AT_EMPTY_PATH, STATX_ALL, &stats[0]);
+	statx(pidfd[1], "", AT_EMPTY_PATH, STATX_ALL, &stats[1]);
+	if (stats[0].stx_ino != stats[1].stx_ino)
+		return 1;
+	return 0;
+}
+
+int main(int argc, char* argv[])
+{
+	#define READ 0
+	#define WRITE 1
+
+	int child, ret, gchild, p[2];
+	int cpidfd[2], gpidfd[2];
+	struct statx stats[2];
+
+	test_init(argc, argv);
+
+	if (pipe(p)) {
+		pr_perror("pipe");
+		return 1;
+	}
+
+	child = test_fork();
+	if (child < 0) {
+		pr_perror("fork");
+		return 1;
+	} else if (child == 0) {
+		int gchild = test_fork();
+		if (gchild < 0) {
+			pr_perror("fork");
+			return 1;
+		} else if (gchild == 0) {
+			close(p[READ]);
+			close(p[WRITE]);
+			while(1);
```

Comment:
do we really need the spin loop here?

Reply:
We would like the grandchild and child to only terminate after the main process has opened pidfds for them successfully. The spin loop ensures that both the child and grandchild are alive until the SIGKILL call on line 148.

Reply:
while (1) sleep(1000) will work either, won't it?

Reply:
Yes, it will! I have updated the test.